### PR TITLE
Refactoring of `pd_gmres` and tests for sanity checks

### PR DESCRIPTION
**Pull request by jhabriel**
This PR introduces further refactoring of the `pd_gmres` function in addition to 18 tests for the sanity checks of the `pd_gmres` function.

The following changes have been introduced:

1. The input argument `m0` was renamed to `mInitial`.
2. The input argument `x0` was renamed to `xInitial`.
3. A new input parameter `mMinMax` controlling the minimum and maximum values of the restart parameter `m` can now be given by the user.
4. A new input parameter `mStep` controlling the step of the restart parameter in the `pd_rule` function can now also be given by the user.
5. The input variables `alphaP` and `alphaD` should now be given as a single input variable `alphaPD`, which is a `2-by-1` vector.
6. The parameter `restarted` is given as an output variable now.

Default values and sanity checks were updated/added accordingly. Documentation on the input parameters was considerably improved. In addition, some cleaning of the code was performed.

In this PR, the test `test_pd_gmres` was renamed to `test_poisson` to better reflect the purpose of the tests.

### CODE REVIEW

**Review by jccf19 — CHANGES_REQUESTED** [review]

File: src/pd_gmres.m [hunk 1]
```diff
@@ -16,20 +24,37 @@
...
-%             Tolerance error threshold for relative residual norm.
-%           
-%   maxit:    int
-%             Maximum number of iterations.
+%   mMinMax:  2-by-1 vector, optional
+%             Minimum and maximum values of the restart paramter m.
+%             Default is [1; n]. Note that  we require 
+%             1 <= mMinMax(1) < mMinMax(2) <= n. Note that for large
+%             matrices (e.g., > 600), the default value mMinMax(2) = n
+%             might result in low convergence. For these cases, following
```
[the remark this answers is not in the document]
Comment (jhabriel) [reply src/pd_gmres.m:38]: Resolved after conversation.

File: src/pd_gmres.m [hunk 2]
```diff
@@ -16,20 +24,37 @@
...
+%             Ref. [2], we recommend something on the order of [1; 100].
```
[the remark this answers is not in the document]
Comment (jhabriel) [reply src/pd_gmres.m:39]: Resolved above.

File: src/pd_gmres.m [hunk 3]
```diff
@@ -16,20 +24,37 @@
...
+%   mStep:    int, optional
+%             Step size for incresing the restart parameter m between
+%             cycles. Default is 1 if n <= 10 and 3 otherwise.
+%
+%   tol:      float, optional
+%             Tolerance error threshold for the relative residual norm.
+%             Default is 1e-6.
+%           
+%   maxit:    int, optional
+%             Maximum number of iterations.
```
[the remark this answers is not in the document]
Comment (jhabriel) [reply src/pd_gmres.m:50]: Incorporated.

File: src/pd_gmres.m [hunk 4]
```diff
@@ -16,20 +24,37 @@
...
+%             Default is [1; n]. Note that  we require 
+%             1 <= mMinMax(1) < mMinMax(2) <= n. Note that for large
+%             matrices (e.g., > 600), the default value mMinMax(2) = n
+%             might result in low convergence. For these cases, following
+%             Ref. [2], we recommend something on the order of [1; 100].
 %
-%   alphaP:   float
-%             Proportional coefficient from the PD controller.
+%   mStep:    int, optional
+%             Step size for incresing the restart parameter m between
```
[the remark this answers is not in the document]
Comment (jhabriel) [reply src/pd_gmres.m:42]: Done!

File: src/pd_gmres.m [hunk 5]
```diff
@@ -97,32 +123,65 @@
...
+if (nargin < 4) || isempty(mMinMax)
+    mMinMax = [1; n];
+else
+    if ~restarted
+        warning("mMinMax was given but will not be used.")
     end
 end
 
+if restarted && ...
+        ((mMinMax(1) < 1) || ...
```
[the remark this answers is not in the document]
Comment (jhabriel) [reply src/pd_gmres.m:157]: Will look into this afterwards.

File: src/pd_gmres.m [hunk 6]
```diff
@@ -196,89 +248,97 @@
...
 
-while flag==0
-    if iter(size(iter,1),:) ~=1
-        [miter]=pd_rule(m,m0,mmin,res,iter(size(iter,1),:),mstep, mmax,alphaP, alphaD);
-        m=miter(1,1);
-        m0=miter(1,2);
+while flag == 0
+    
+    if iter(size(iter, 1), :) ~= 1
+        [miter] = pd_rule(m, mInitial, mMin, res, ...
```
[the remark this answers is not in the document]
Comment (jhabriel) [reply src/pd_gmres.m:264]: Same as above.

File: src/pd_gmres.m [hunk 7]
```diff
@@ -16,20 +24,37 @@
...
+%             Maximum number of iterations. 
 %
-%   alphaD:   float
-%             Derivative coefficient from the PD controller.
+%   xInitial: n-by-1 vector, optional
+%             Vector of initial guess. Default is zeros(n, 1).
+% 
+%   alphaPD:  2-by-1 vector, optional
+%             Proportional and derivative coefficients from the
+%             proportional-derivative controller. Default is [-5, 3].
```
[the remark this answers is not in the document]
Comment (jhabriel) [reply src/pd_gmres.m]: Fixed.

**Review by jccf19 — APPROVED** [review]
Thanks!

**Pull request merged**